server/lib/nl/quantity_parser.py

```diff
@@ -0,0 +1,188 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Helper for parsing quantity"""
+
+# This library supports parsing numeric values and ranges from a query.
+# It requires numbers plus an optional multipler in words, like
+# 1M, 10 million, 2 trillion, 30K.
+
+# TODO: support non-digits too, like "one thousand".
+# TODO: maybe support units
+
+import logging
+import re
+
+from server.lib.nl.counters import Counters
+from server.lib.nl.detection import QCmpType
+from server.lib.nl.detection import Quantity
+from server.lib.nl.detection import QuantityClassificationAttributes
+from server.lib.nl.detection import QuantityRange
+
+# Match a number followed by an optional decimal part.
+NUMBER_RE = r'\d+(?:\.\d+)?'
+
+# Match a bunch of words indicating multipliers to be applied to a number.
+NUMBER_FACTOR_RE = r'k|m|b|t|hundred|thousand|million|billion|trillion'
+
+SPACE_RE = r'(?: +)?'
+
+NUMBER_FACTOR_MAP = {
+    'hundred': 100,
+    'k': 1000,
+    'thousand': 1000,
+    'm': 1000000,
+    'million': 1000000,
+    'b': 1000000000,
+    'billion': 1000000000,
+    't': 1000000000000,
+    'trillion': 1000000000000,
+}
+
+
+# Match x at word boundary.
+def _matchable(x):
+  return r'(?:^|\W)' + x + r'(?:$|\W)'
+
+
+# Allow a bunch of stuff between X and number. Followed by an optional space and an optional factor.
+def _digits(x):
+  # Capture number space and factor all together as a single block.
+  return x + r'\s*?(' + NUMBER_RE + SPACE_RE + r'(?:' + NUMBER_FACTOR_RE + r')?)'
+
+
+# The order here matters. The ones earlier in the list may include
+# sub-strings that appear later.
+QUANTITY_RE = [
+    (QCmpType.GE,
+     _digits(r'(?:>=|greater than or equal to|greater than or equals|from)')),
+    (QCmpType.LE,
+     _digits(
+         r'(?:<=|less than or equal to|less than or equals|lesser than or equal to|lesser than or equals|upto|up to)'
+     )),
+    (QCmpType.GT,
+     _digits(
+         r'(?:>|over|above|exceed|exceeds|more than|greater than|higher than|upwards of)'
+     )),
+    (QCmpType.LT,
+     _digits(r'(?:<|under|below|less than|lesser than|lower than)')),
+    (QCmpType.EQ, _digits(r'(?:==?|is|equal to|equals|has|have|as much as)'))
```

Comment:
nit: extra space between "equal" and "to"?

Reply:
I think there was only 1 space, and validated in this unit test check - https://github.com/datacommonsorg/website/blob/1b1a045721bf72401ff4854bdb5a9f8f5e5a245a/server/tests/lib/nl/quantity_parser_test.py#L32